### PR TITLE
Fix segment map modification in OOM scenarios

### LINKED ISSUE #9064: GC/Coverage/delete_next_card_table/delete_next_card_table failure in CI
From https://ci.dot.net/job/dotnet_coreclr/job/master/job/debug_windows_nt_prtest/2004/console:

```
10:43:44 D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\TestWrappers\GC\GC.XUnitWrapper.cs(4866): error : GC._Coverage_delete_next_card_table_delete_next_card_table_._Coverage_delete_next_card_table_delete_next_card_table_cmd [FAIL] [D:\j\workspace\debug_windows---17180f1b\tests\runtest.proj]
10:43:44         
10:43:44         Assert failure(PID 4168 [0x00001048], Thread: 5604 [0x15e4]): !CREATE_CHECK_STRING(!"Detected use of a corrupted OBJECTREF. Possible GC hole.")
10:43:44         
10:43:44         CORECLR! `Object::ValidateInner'::`1'::catch$5 + 0xE9 (0x00007ffa`401c91ce)
10:43:44         CORECLR! CallSettingFrame + 0x20 (0x00007ffa`40014c10)
10:43:44         CORECLR! _CxxCallCatchBlock + 0x15A (0x00007ffa`4000f3ea)
10:43:44         NTDLL! RtlCaptureContext + 0x3C3 (0x00007ffa`56052a63)
10:43:44         CORECLR! Object::ValidateInner + 0x10F (0x00007ffa`3f465b4f)
10:43:44         CORECLR! Object::Validate + 0xE8 (0x00007ffa`3f465488)
10:43:44         CORECLR! WKS::gc_heap::background_promote + 0x117 (0x00007ffa`3fdc7437)
10:43:44         CORECLR! PinObject + 0x195 (0x00007ffa`3fda4e45)
10:43:44         CORECLR! ScanConsecutiveHandlesWithoutUserData + 0xB3 (0x00007ffa`3fe33be3)
10:43:44         CORECLR! BlockScanBlocksWithoutUserData + 0x5C (0x00007ffa`3fe3301c)
10:43:44             File: d:\j\workspace\debug_windows---17180f1b\src\vm\object.cpp Line: 1827
10:43:44             Image: D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\CoreRun.exe
10:43:44         
10:43:44         
10:43:44   
10:43:44   Return code:      1
10:43:44   Raw output file:      D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\Reports\GC\Coverage\delete_next_card_table\delete_next_card_table.output.txt
10:43:44   Raw output:
10:43:44   BEGIN EXECUTION
10:43:44          "D:\j\workspace\debug_windows---17180f1b\tests\..\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe" delete_next_card_table.exe 
10:43:44         Churn loop 0...............................Expected: 100
10:43:44         Actual: 123456789
10:43:44         END EXECUTION - FAILED
10:43:44         FAILED
10:43:44         Test Harness Exitcode is : 1
10:43:44         
10:43:44   To run the test:
10:43:44   > set CORE_ROOT=D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root
10:43:44   > D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\GC\Coverage\delete_next_card_table\delete_next_card_table.cmd
10:43:44   
10:43:44         Expected: True
10:43:44         Actual:   False
10:43:44           D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\TestWrappers\GC\GC.XUnitWrapper.cs(4866,0): at GC._Coverage_delete_next_card_table_delete_next_card_table_._Coverage_delete_next_card_table_delete_next_card_table_cmd()
```
See https://github.com/dotnet/coreclr/pull/8910 for an occurrence.

### PR DESCRIPTION
**Pull request by swgillespie**
Fix an issue where the segment mapping table is not reverted back to its original state when failing to commit the mark array in an OOM scenario. Fixes #9064. The repro that I have does not reproduce any longer with this change in place.

cc @Maoni0 @sergiy-k @adityamandaleeka

### CODE REVIEW

**Comment by Maoni0** [comment]
LGTM

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -7202,14 +7203,18 @@ int gc_heap::grow_brick_card_tables (uint8_t* start,
...
-            seg_mapping* new_seg_mapping_table = (seg_mapping*)(mem + st_table_offset_aligned);
+            new_seg_mapping_table = (seg_mapping*)(mem + st_table_offset_aligned);
             new_seg_mapping_table = (seg_mapping*)((uint8_t*)new_seg_mapping_table -
                                               size_seg_mapping_table_of (0, (align_lower_segment (saved_g_lowest_address))));
             memcpy(&new_seg_mapping_table[seg_mapping_word_of(g_gc_lowest_address)],
                 &seg_mapping_table[seg_mapping_word_of(g_gc_lowest_address)],
                 size_seg_mapping_table_of(g_gc_lowest_address, g_gc_highest_address));
 
-            seg_mapping_table = new_seg_mapping_table;
+            // new_seg_mapping_table gets assigned to seg_mapping_table at the bottom of this function,
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: There's a dprintf statement between here and where seg_mapping_table is being set that prints the seg_mapping_table. Maybe you should print the new one there? Or call out that the one being printed is the prior one.

**Comment by adityamandaleeka** [comment]
LGTM apart from the minor comment above.

**Comment by swgillespie** [comment]
@dotnet-bot test OSX Release longgc
@dotnet-bot test Ubuntu Release longgc
@dotnet-bot test Windows_NT Release longgc
@dotnet-bot test OSX Release gcsimulator

**Comment by Maoni0** [comment]
actually I thought about this a bit more and you should update the seg_mapping_table before updating the new g_gc* addresses. people could be calling this in non heap pointer which could land in the new range so you need to have the new table ready before the range is updated. sorry about that.

**Comment by swgillespie** [comment]
no problem, done!

**Comment by swgillespie** [comment]
thanks!

**Pull request merged**